exla/lib/exla/defn.ex

```diff
@@ -546,6 +558,48 @@ defmodule EXLA.Defn do
     end
   end
 
+  defp cached_recur_operator(
+         :elixir_call,
+         %T{data: %Expr{args: [tensor_expr, opts, fun, out_template]}} = expr,
+         %{client: %EXLA.Client{platform: :host}, callback_server_pid: callback_server_pid} = state,
+         cache
+       ) do
+    # Flatten the tensor_or_container expression into its tensor leaves so we
+    # can compile each as an independent operand to the host callback.
+    tensor_exprs = Composite.flatten_list([tensor_expr])
+
+    {arg_values, cache} =
+      Enum.map_reduce(tensor_exprs, cache, fn arg, cache ->
+        recur_operator(arg, state, cache) |> unwrap_single_tensor!()
+      end)
+
+    # Build a template container for the tensor_or_container argument so the
+    # callback server can reconstruct the full structure from a flat list of
+    # decoded tensors.
+    arg_template = Nx.to_template(tensor_expr)
+
+    callback_id = EXLA.CallbackServer.register(callback_server_pid, fun, out_template, arg_template, opts)
```

Comment:
I wonder if should rely on Expr ID to use it as a key instead of using all of the arguments.